cmd/attest.go

```diff
@@ -0,0 +1,267 @@
+    {{.appName}} {{.command}} oci-dir:path/to/yourimage              read directly from a path on disk for OCI layout directories (from Skopeo or otherwise)
+    {{.appName}} {{.command}} registry:yourrepo/yourimage:tag        pull image directly from a registry (no container runtime required)
+`
+)
+
+var (
+	keyPath           string
+	attestationOutput []string
+	validAttestScheme = []string{
+		"docker",
+		"docker-archive",
+		"oci-archive",
+		"oci-dir",
+		"registry",
+	}
+	attestCmd = &cobra.Command{
+		Use:   "attest --output [FORMAT] --key [KEY] [SOURCE]",
+		Short: "Generate a package SBOM as an attestation to [SOURCE]",
+		Long:  "Generate a packaged-based Software Bill Of Materials (SBOM) from a container image or OCI directory as the predicate of an attestation.",
+		Example: internal.Tprintf(attestExample, map[string]interface{}{
+			"appName": internal.ApplicationName,
+			"command": "attest",
+		}),
+		Args:          validateInputArgs,
+		SilenceUsage:  true,
+		SilenceErrors: true,
+		PreRunE: func(cmd *cobra.Command, args []string) (err error) {
+			if appConfig.Dev.ProfileCPU && appConfig.Dev.ProfileMem {
+				return fmt.Errorf("cannot profile CPU and memory simultaneously")
+			}
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if appConfig.Dev.ProfileCPU {
+				defer profile.Start(profile.CPUProfile).Stop()
+			} else if appConfig.Dev.ProfileMem {
+				defer profile.Start(profile.MemProfile).Stop()
+			}
+
+			return attestExec(cmd.Context(), cmd, args)
+		},
+	}
+)
+
+func isTerminal() bool {
+	stat, _ := os.Stdin.Stat()
+	return (stat.Mode() & os.ModeCharDevice) != 0
+}
+
+func getPassFromTerm(confirm bool) ([]byte, error) {
+	fmt.Fprint(os.Stderr, "Enter password for private key: ")
+	pw1, err := term.ReadPassword(0)
+	if err != nil {
+		return nil, err
+	}
+	if !confirm {
+		return pw1, nil
+	}
+	fmt.Fprintln(os.Stderr)
+	fmt.Fprint(os.Stderr, "Enter password for private key again: ")
+	confirmpw, err := term.ReadPassword(0)
+	fmt.Fprintln(os.Stderr)
+	if err != nil {
+		return nil, err
+	}
+
+	if string(pw1) != string(confirmpw) {
+		return nil, errors.New("passwords do not match")
+	}
+	return pw1, nil
+}
+
+func passFunc(isPass bool) (b []byte, err error) {
+	pw, ok := os.LookupEnv("COSIGN_PASSWORD")
+	switch {
+	case ok:
+		return []byte(pw), nil
+	case isTerminal():
+		return getPassFromTerm(false)
+	// Handle piped in passwords.
+	default:
+		return io.ReadAll(os.Stdin)
+	}
+}
+
+func validateScheme(userInput string) error {
+	switch {
+	case strings.HasPrefix(userInput, "dir"), strings.HasPrefix(userInput, "file"):
+		return fmt.Errorf("could not support attestation for %s; please try one of these scheme: %v", userInput, validAttestScheme)
+	default:
+		return nil
+	}
+}
+
+func attestExec(ctx context.Context, _ *cobra.Command, args []string) error {
+	// can only be an image for attestation or OCI DIR
+	userInput := args[0]
+	err := validateScheme(userInput)
```

Comment:
this validation only partially covers all of the source type cases. `generateSBOM` can still create SBOMs for sources that are files or dirs:
```
syft attest .
```

We can export `detectScheme` so that something before `generateSBOM` is called to help with this (https://github.com/anchore/syft/blob/5debdcbc9d2cffda6d15f11ae2437a163bdf0e78/syft/source/source.go#L41)